pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/model/BuildCondition.java

```diff
@@ -130,8 +140,28 @@ protected Result getExecutionResult(@Nonnull WorkflowRun r) {
         return conditions;
     }
 
+    @Deprecated
     @Nonnull
     public static Result getCombinedResult(@Nonnull WorkflowRun run, @CheckForNull Throwable error) {
+        return getCombinedResult(run, error, null, null);
+    }
+
+    @Nonnull
+    public static Result getCombinedResult(@Nonnull WorkflowRun run, @CheckForNull Throwable error, @CheckForNull FlowNode startNode, @CheckForNull FlowNode endNode) {
+        if (startNode != null && endNode != null) {
+            // TODO: Remove this in favor of moving StatusAndTiming#findWarningBetween public.
+            DepthFirstScanner scanner = new DepthFirstScanner();
+            if (scanner.setup(endNode, Collections.singletonList(startNode))) {
+                WarningAction warningAction = StreamSupport.stream(scanner.spliterator(), false)
+                        .map(node -> node.getPersistentAction(WarningAction.class))
+                        .filter(Objects::nonNull)
+                        .max(Comparator.comparing(warning -> warning.getResult().ordinal))
+                        .orElse(null);
+                if (warningAction != null) {
+                    return warningAction.getResult();
```

Comment:
I think you don't want an explicit return here, because you want to use the worst of the stage, exec, run, and error results, right?

Reply:
Do we? I'm legitimately not sure.

Reply:
That is to say - do we want to have a successful individual stage trigger `post` `success` even if the overall build status is failure? I feel like we will...

Reply:
Hmm. My concern is breaking existing use cases where people are setting the build result explicitly, but now that I'm thinking about it it might be more nuanced in this case.

Here is the kind of thing where I'm not sure what makes the most sense:

```
pipeline {
    agent any

     stages {
        stage("foo") {
            steps {
                // Default Settings
                catchError(buildResult: "FAILURE", stageResult: "SUCCESS") {
                    error("uhoh")
                }
                catchError(buildResult: "SUCCESS", stageResult: "UNSTABLE") {
                    error("uhoh")
                }
            }
            post {
                unstable {
                    echo "Does this run?"
                }
                failure {
                    echo "Or this?"
                }
            }
        }
    }
}
```
Anyone using `catchError(buildResult: 'SUCCESS', stageResult: 'blah')` is probably going to be careful to not set the build result anywhere intentionally, so maybe the current behavior is the best option?

Reply:
> do we want to have a successful individual stage trigger post success even if the overall build status is failure?

`catchError` does not create a `WarningAction` for `SUCCESS`, so it wouldn't hit the return and we'd use the overall build result. (Although we could change that behavior if it would make more sense for the use case here since it wouldn't break anything else).

Reply:
Ok, yeah, let's just fallback then.